File: ios/Framework/Core/DDL/ListScreenlet/DDLListScreenlet.swift
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */
import UIKit


@objc public protocol DDLListScreenletDelegate : BaseScreenletDelegate {

	@objc optional func screenlet(_ screenlet: DDLListScreenlet,
			onDDLListResponseRecords records: [DDLRecord])

	@objc optional func screenlet(_ screenlet: DDLListScreenlet,
			onDDLListError error: NSError)

	@objc optional func screenlet(_ screenlet: DDLListScreenlet,
			onDDLSelectedRecord record: DDLRecord)

}


open class DDLListScreenlet: BaseListScreenlet {


	//MARK: Inspectables
	
	@IBInspectable open var userId: Int64 = 0

	@IBInspectable open var recordSetId: Int64 = 0

	@IBInspectable open var labelFields: String? {
		didSet {
			(screenletView as? DDLListViewModel)?.labelFields = parseFields(labelFields)
		}
	}

	@IBInspectable open var offlinePolicy: String? = CacheStrategyType.remoteFirst.rawValue


	open var ddlListDelegate: DDLListScreenletDelegate? {
		return delegate as? DDLListScreenletDelegate
	}

	open var viewModel: DDLListViewModel {
		return screenletView as! DDLListViewModel
	}


	//MARK: BaseListScreenlet

	override open func onCreated() {
		super.onCreated()

		viewModel.labelFields = parseFields(self.labelFields)
	}

	override open func createPageLoadInteractor(
			page: Int,
			computeRowCount: Bool)
			-> BaseListPageLoadInteractor {

		let interactor = DDLListPageLoadInteractor(
				screenlet: self,
				page: page,
				computeRowCount: computeRowCount,
				userId: self.userId,
				recordSetId: self.recordSetId)

		interactor.cacheStrategy = CacheStrategyType(rawValue: self.offlinePolicy ?? "") ?? .remoteFirst

		return interactor
	}

	override open func onLoadPageError(page: Int, error: NSError) {
		super.onLoadPageError(page: page, error: error)

		ddlListDelegate?.screenlet?(self, onDDLListError: error)
	}

	override open func onLoadPageResult(page: Int, rows: [AnyObject], rowCount: Int) {
		super.onLoadPageResult(page: page, rows: rows, rowCount: rowCount)

		ddlListDelegate?.screenlet?(self,
				onDDLListResponseRecords: rows as! [DDLRecord])
	}

	override open func onSelectedRow(_ row: AnyObject) {
		ddlListDelegate?.screenlet?(self,
				onDDLSelectedRecord: row as! DDLRecord)
	}


	//MARK: Private methods

	fileprivate func parseFields(_ fields: String?) -> [String] {
		var result: [String] = []

		if let fieldsValue = fields {
			let dirtyFields = (fieldsValue as NSString).components(separatedBy: ",")
			result = dirtyFields.map() {
				$0.trimmingCharacters(in: .whitespaces)
			}
			result = result.filter() { return $0 != "" }
		}

		return result
	}

}

File: android/library/src/main/java/com/liferay/mobile/screens/ddl/list/DDLListScreenlet.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.mobile.screens.ddl.list;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import com.liferay.mobile.screens.R;
import com.liferay.mobile.screens.base.list.BaseListScreenlet;
import com.liferay.mobile.screens.base.list.interactor.BaseListInteractorListener;
import com.liferay.mobile.screens.ddl.list.interactor.DDLListInteractor;
import com.liferay.mobile.screens.ddl.model.Record;

/**
 * @author Javier Gamarra
 * @author Silvio Santos
 */
public class DDLListScreenlet extends BaseListScreenlet<Record, DDLListInteractor>
    implements BaseListInteractorListener<Record> {

    private long recordSetId;

    public DDLListScreenlet(Context context) {
        super(context);
    }

    public DDLListScreenlet(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public DDLListScreenlet(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public DDLListScreenlet(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public long getRecordSetId() {
        return recordSetId;
    }

    public void setRecordSetId(long recordSetId) {
        this.recordSetId = recordSetId;
    }

    @Override
    public void error(Exception e, String userAction) {
        if (getListener() != null) {
            getListener().error(e, userAction);
        }
    }

    @Override
    protected void loadRows(DDLListInteractor interactor) {
        interactor.start(recordSetId, userId);
    }

    @Override
    protected View createScreenletView(Context context, AttributeSet attributes) {
        TypedArray typedArray =
            context.getTheme().obtainStyledAttributes(attributes, R.styleable.DDLListScreenlet, 0, 0);
        recordSetId = castToLong(typedArray.getString(R.styleable.DDLListScreenlet_recordSetId));
        typedArray.recycle();

        return super.createScreenletView(context, attributes);
    }

    @Override
    protected DDLListInteractor createInteractor(String actionName) {
        return new DDLListInteractor();
    }
}